segmentation/utils/phenotype_spatial_analysis.py

```diff
@@ -0,0 +1,106 @@
+
+    Args:
+        z: array of z scores of shape (pheno_num x pheno_num)
+        pheno_titles: list of all the names of the phenotypes"""
+    # visualize
+    zplot = np.array(z)
+    # Replace the NA's and inf values with 0s
+    zplot[np.isnan(zplot)] = 0
+    zplot[np.isinf(zplot)] = 0
+    # Assign numpy values respective phenotype labels
+    zplot = pd.DataFrame(zplot, columns=pheno_titles, index=pheno_titles)
+    sns.set(font_scale=.7)
+    sns.clustermap(zplot, figsize=(8, 8), cmap="vlag")
+
+
+def calculate_phenotype_spatial_enrichment(all_patient_data, pheno, dist_mat, points=None,
+                                           patient_idx=0, cell_label_idx=1, flowsom_idx=52,
+                                           bootstrap_num=1000, dist_lim=100):
+    """Spatial enrichment analysis based on cell phenotypes to find significant interactions between different
+    cell types, looking for both positive and negative enrichment. Uses bootstrapping to permute cell labels randomly.
+
+    Args:
+        all_patient_data: data including points, cell labels, and
+            cell expression matrix for all markers
+        pheno: array containing the cell phenotypes in the population and their corresponding IDs
+        dist_mat: cells x cells matrix with the euclidian
+            distance between centers of corresponding cells
+        points: patient labels to include in analysis. If argument is none, default is all labels used
+        patient_idx: columns with patient labels. Default is 0
+        cell_label_idx: column with cell labels. Default is 1
+        flowsom_idx: column with the phenotype IDs. Default is 52
+        bootstrap_num: number of permutations for bootstrap. Default is 1000
+        dist_lim: cell proximity threshold. Default is 100
+
+    Returns:
+        values: a list with each element consisting of a tuple of
+            closenum and closenumrand for each point included in the analysis
+        stats: an Xarray with dimensions (points, stats, number of markers, number of markers) The included stats
+            variables are:
+            z, muhat, sigmahat, p, h, adj_p, and marker_titles for each point in the analysis"""
+
+    # Setup input and parameters
+    num_points = 0
+    if points is None:
+        points = list(set(all_patient_data.iloc[:, patient_idx]))
+        num_points = len(points)
+    else:
+        num_points = len(points)
+
+    cell_pheno_idx = 2
+    values = []
+
+    # Error Checking
+    if not np.isin(points, all_patient_data.iloc[:, patient_idx]).all():
+        raise ValueError("Points were not found in Expression Matrix")
+
+    # Subset matrix to only include the columns with the patient label, cell label, and cell phenotype
+    all_patient_data = all_patient_data[all_patient_data.columns[[patient_idx, cell_label_idx, flowsom_idx]]]
+    # Extract the names of the cell phenotypes
+    pheno_titles = pheno.iloc[:, 0]
```

Comment:
The separate pheno names is an artifact of matlab, which doesn't let you have mixed numeric and string data. We can have one column in patient_data that has the number, and a second column in patient_data that has the name. We don't need to store this separately. This will simplify things a bit, right now we have a lot of different indexes, which is a bit confusing